README.md
# MRDatabaseContentChecker

[![Version](http://cocoapod-badges.herokuapp.com/v/MRDatabaseContentChecker/badge.png)](http://cocoadocs.org/docsets/MRDatabaseContentChecker)
[![Platform](http://cocoapod-badges.herokuapp.com/p/MRDatabaseContentChecker/badge.png)](http://cocoadocs.org/docsets/MRDatabaseContentChecker)

MRDatabaseContentChecker is a small library to make checking the contents
of a database easier and more literate in your tests. It tries to stay out
your way and make writing tedious checks simpler and less error prone.

You can test the content of either tables or queries.

MRDatabaseContentChecker requires that you are using FMDB.

## Usage

You can check either the full content of database tables, or the results of
a query (as they're both table-like structures).

### Tables

The simplest example, using SenTest, is something like the following:

```objc
FMDatabase *db = [...];
MRDatabaseContentChecker *dc = [[MRDatabaseContentChecker alloc] init];

NSError *validationError;
NSArray *expectedRows = @[
    @[@"first_name", @"surname"],  // first array is column headers
    @[@"Mike",       @"Rhodes"],   // expected data starts in row 2
    @[@"John",       @"Smith"]     // and continues...
];
STAssertTrue([dc checkDatabase:db
                         table:@"users"
                       hasRows:expectedRows
                         error:&validationError],
             [dc formattedErrors:validationError]);
```

Essentially:

1. Create your expected values array. This is an array of arrays. The first
   array always contains the column headers you want to check. The following
   rows define the data you expect to find.
1. Call `-checkDatabase:table:hasRows:error:` in a testing macro.
1. If that returns `NO`, use the `-formattedErrors` helper to print the
   failures.

### Queries

The above example can be rewritten to use a query rather than a table name:

```objc
FMDatabase *db = [...];
MRDatabaseContentChecker *dc = [[MRDatabaseContentChecker alloc] init];

NSError *validationError;
NSArray *expectedRows = @[  /* as before */  ];
STAssertTrue([dc checkDatabase:db
                         query:@"select first_name, surname from users;"
                       hasRows:expectedRows
                         error:&validationError],
             [dc formattedErrors:validationError]);
```

At the moment, there's no support for placeholders; you control the SQL
you're generating in your tests. Or, if not, feel free to open a PR as I'm
not in the slightest averse to including further features.

### Caveats

* The table-checking variant relies on the "undefined" ordering that SQLite
  returns rows in. Right now, if this doesn't work, using the query variant is
  neccessary to allow an `ORDER BY` clause to be included.
* There is no limit enforced on the number of errors reported. This may be
  a problem if there are millions of errors as every error has an
  entry in the error object passed out of the method.

### Checks reference

`-checkDatabase:table:hasRows:error:` and `-checkDatabase:query:hasRows:error:`
check:

- Each expected value matches is corresponding item in the result set.
- There are the same number of results as expected.

#### Types

As shown above, expected values are passed as an array of arrays. Each
expected value is check against its corresponding entry in the result set
based on its type. The allowed types are:

<dl>
<dt>NSString</dt>
<dd>NSString objects are checked using -isEqual.</dd>
<dt>NSNumber</dt>
<dd>NSNumber objects are checked using -isEqual against -longLongIntForColumn
values from the database.</dd>
<dt>BOOL (boxed)</dt>
<dd>As a boxed BOOL is an NSNumber, see NSNumber.</dd>
<dt>NSData</dt>
<dd>NSData objects are checked using -isEqual against -dataForColumn
value from database.</dd>
<dt>NSRegularExpression</dt>
<dd>This check passes if the return value from
-numberOfMatchesInString:options:error: returns one or more matches.</dd>
<dt>[NSNull null]</dt>
<dd>This check checks [NSNull null] against the value returned for
the column by -objectForColumnName:.</dd>
</dl>

Using an unsupported type (or a class that isn't a subclass of a supported
type) will result in a failure checking the value.

You can pass different datatypes for a given column's expected values:

```objc
expectedRows = @[
    @[@"name",        @"github_username", @"age"],
    @[@"Mike Rhodes", @"mikerhodes",      @30],
    @[@"John Smith",  [NSNull null],      @27]
];
```

## Requirements

MRDatabaseContentChecker depends on:

* FMDB

## Installation

MRDatabaseContentChecker is available through [CocoaPods](http://cocoapods.org),
to install it add the following line to your Podfile:

    pod "MRDatabaseContentChecker"

## Author

Michael Rhodes, mike.rhodes@gmail.com

## License

MRDatabaseContentChecker is available under the Apache v2 license. See the LICENSE file for more info.



CHANGELOG.md
# MRDatabaseContentChecker CHANGELOG

## 0.4.0 (2016-01-11)

- Support building as modular framework.

## 0.3.0

- Allow ordering results for a db query during checking.

## 0.2.0

Unsure on changes

## 0.1.0

Initial release.


Classes/DatabaseContentChecker.h
//
//  DatabaseContentChecker.h
//
//  Created by Michael Rhodes on 11/05/2014.
//  Copyright (c) 2014 Mike Rhodes. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

#import <Foundation/Foundation.h>

#import <FMDatabase.h>
#import <FMDatabaseQueue.h>


/**
 * Indexing and query erors.
 */
typedef NS_ENUM(NSInteger, DatabaseContentCheckerError) {
    /**
     Validation of database failed
     */
    DatabaseContentCheckerErrorValidation = 1
};

/**
 Error domain for the database checker
 */
extern NSString* const DatabaseContentCheckerErrorDomain;

/**
 Key in user info dict for error messages.

 A typical use would be to use componentsJoinedByString: to
 join in an assert message.
 */
extern NSString* const DatabaseContentCheckerErrorsArray;

@interface DatabaseContentChecker : NSObject

- (NSString*)formattedErrors:(NSError*)error;

- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedContent
                error:(NSError* __autoreleasing *)error;

@end


Classes/DatabaseContentChecker.m
//
//  DatabaseContentChecker.m
//
//  Created by Michael Rhodes on 11/05/2014.
//  Copyright (c) 2014 Mike Rhodes. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

#import "DatabaseContentChecker.h"

NSString* const DatabaseContentCheckerErrorDomain =
@"DatabaseContentCheckerErrorDomain";
NSString* const DatabaseContentCheckerErrorsArray =
@"DatabaseContentCheckerErrorsArray";

/**
 A simple class to check the content of your database tables.
 */
@implementation DatabaseContentChecker

/**
 Format the error messages produced by other methods in this
 class.
 */
- (NSString*)formattedErrors:(NSError*)error
{
    NSArray *errors = error.userInfo[DatabaseContentCheckerErrorsArray];
    return [errors componentsJoinedByString:@"; "];
}

/**
 Check that the contents of `table` match the specification in
 `expectedRows`. The first array in `expectedRows` MUST be the
 column names you want to check.

 Rows are assumed to be in the order presented in the array, and the table
 is assumed to contain the same number of rows as in the array.

 Note this means you can check just a subset of the data in a
 given table by including a subset of the columns.

 For more information on what you can have as values to check, see
 the documentation for checkDatabase:query:hasRows:error:, as this
 method calls into that one after constructing a query to get all
 the rows on the table.
 */
- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedRows
                error:(NSError* __autoreleasing *)error
{
    NSArray *columns = expectedRows[0];

    NSString *sql = [NSString stringWithFormat:@"select %@ from %@",
                     [columns componentsJoinedByString:@", "],
                     table];

    return [self checkDatabase:db
                         query:sql
                       hasRows:expectedRows
                         error:error];
}

/**
 Check that the result of `query` matches the specification in
 `expectedRows`. The first array in `expectedRows` MUST be the
 column names you want to check from the result set. For example:

 ```
 NSArray *expectedRows = @[
 @[@"docId",      @"revId",  @"sequenceNumber"],  // columns
 @[@"document-1", @"1-rev",  @(1)],               // expected values
 @[@"document-2", @"34-rev", @(2)]
 ];
 ```

 Expected values can be: NSString, NSNumber, NSData or NSRegularExpression.

 In the case of regular expressions, checking is based on the existence
 of one or more matches using numberOfMatchesInString:options:error:.

 Rows are assumed to be present in the result set in the order presented
 in the array, and the result set is assumed to have the same number of
 results as the array. This means that you may want to use order by or
 limit clauses in your query.

 Note this means you can check just a subset of the data in a
 given query by including a subset of the columns.
 */
- (BOOL)checkDatabase:(FMDatabase*)db
                query:(NSString*)sql
              hasRows:(NSArray*)expectedRows
                error:(NSError* __autoreleasing *)error
{
    NSArray *columns = expectedRows[0];
    NSRange range = NSMakeRange(1, [expectedRows count] - 1);
    NSArray *values = [expectedRows subarrayWithRange:range];

    FMResultSet *result = [db executeQuery:sql];

    BOOL success = YES;
    NSMutableArray *errors = [[NSMutableArray alloc] init];

    for (int i = 0; i < values.count; i++) {
        NSArray *expectedRow = values[i];

        if (![result next]) {
            success = NO;
            NSString *msg = @"Reached end of result set before array";
            [errors addObject:msg];
            break;
        }

        NSLog(@"testing content of docs table");

        for (int col = 0; col < columns.count; col++) {
            NSString *columnName = columns[col];
            NSObject *expectedValue = expectedRow[col];

            BOOL correct;
            Class expectedClass = [expectedValue class];
            NSObject *actualValue;

            if ([expectedClass isSubclassOfClass:[NSString class]]) {
                actualValue = [result stringForColumn:columnName];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSNumber class]]) {
                long long v = [result longLongIntForColumn:columnName];
                actualValue = [NSNumber numberWithLongLong:v];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSData class]]) {
                actualValue = [result dataForColumn:columnName];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSRegularExpression class]]) {
                actualValue = [result stringForColumn:columnName];
                NSString *actualString = (NSString*)actualValue;
                NSRange range = NSMakeRange(0, actualString.length);
                NSRegularExpression *expectedRegex = (NSRegularExpression*)expectedValue;
                NSUInteger found = [expectedRegex numberOfMatchesInString:actualString
                                                                  options:0
                                                                    range:range];
                correct = (found > 0);

            } else if (expectedValue == [NSNull null]) {
                actualValue = [result objectForColumnName:columnName];
                correct = actualValue == [NSNull null];

            } else {
                // Unknown class type!
                correct = NO;
            }

            if (!correct) {
                success = NO;

                NSString *msg = [NSString stringWithFormat:
                                 @"Row %i, field %@, expected %@, actual %@",
                                 i,
                                 columnName,
                                 expectedValue,
                                 actualValue];
                [errors addObject:msg];
            }
        }
    }

    if ([result next]) {  // we should be at the end of the results
        success = NO;

        NSString *msg = @"End of array not end of result set";
        [errors addObject:msg];
    }

    [result close];

    if (errors.count > 0 && error) {
        NSString *description = NSLocalizedString(
                                                  @"Problem updating attachments table.",
                                                  nil);
        NSDictionary *userInfo = @{
                                   NSLocalizedDescriptionKey: description,
                                   DatabaseContentCheckerErrorsArray: [errors copy]
                                   };
        *error = [NSError errorWithDomain:DatabaseContentCheckerErrorDomain
                                     code:DatabaseContentCheckerErrorValidation
                                 userInfo:userInfo];
    }

    return success;
}

@end


Classes/MRDatabaseContentChecker.h
//
//  MRDatabaseContentChecker.h
//
//  Created by Michael Rhodes on 11/05/2014.
//  Copyright (c) 2014 Mike Rhodes. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

#import <Foundation/Foundation.h>

#import <FMDB/FMDatabase.h>
#import <FMDB/FMDatabaseQueue.h>


/**
 * Indexing and query erors.
 */
typedef NS_ENUM(NSInteger, MRDatabaseContentCheckerError) {
    /**
     Validation of database failed
     */
    MRDatabaseContentCheckerErrorValidation = 1
};

/**
 Error domain for the database checker
 */
extern NSString* const MRDatabaseContentCheckerErrorDomain;

/**
 Key in user info dict for error messages.

 A typical use would be to use componentsJoinedByString: to
 join in an assert message.
 */
extern NSString* const MRDatabaseContentCheckerErrorsArray;

@interface MRDatabaseContentChecker : NSObject

- (NSString*)formattedErrors:(NSError*)error;

- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedContent
                error:(NSError* __autoreleasing *)error;

- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedContent
              orderBy:(NSArray*)orderby
                error:(NSError* __autoreleasing *)error;

@end


Example/DatabaseContentCheckerExample/DCCAppDelegate.m
//
//  DCCAppDelegate.m
//  DatabaseContentCheckerExample
//
//  Created by Michael Rhodes on 12/05/2014.
//  Copyright (c) 2014 Small Text. All rights reserved.
//

#import "DCCAppDelegate.h"

@implementation DCCAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    self.window = [[UIWindow alloc] initWithFrame:[[UIScreen mainScreen] bounds]];
    // Override point for customization after application launch.
    self.window.backgroundColor = [UIColor whiteColor];
    [self.window makeKeyAndVisible];
    return YES;
}

- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later. 
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Example/DatabaseContentCheckerExampleTests/DatabaseContentCheckerExampleTests.m
//
//  DatabaseContentCheckerExampleTests.m
//  DatabaseContentCheckerExampleTests
//
//  Created by Michael Rhodes on 12/05/2014.
//  Copyright (c) 2014 Small Text. All rights reserved.
//

#import <XCTest/XCTest.h>
#import "FMDatabase.h"
#import "MRDatabaseContentChecker.h"

@interface DatabaseContentCheckerExampleTests : XCTestCase
@property FMDatabase *db;
@end

@implementation DatabaseContentCheckerExampleTests

- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
    
    //need to create a databse
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString *cachePath = [paths objectAtIndex:0];
    NSString *dbPath = [cachePath stringByAppendingPathComponent:@"database.sqlite"];
    
    self.db = [FMDatabase databaseWithPath:dbPath];
    [self.db open];
    [self.db executeUpdate:@"CREATE TABLE test(textData text, numericData int, moreText text)"];
    
    // Building the string ourself
    NSArray * array = @[@4, @1, @9];
    
    for(int i=0; i<3; i++) {
    
        NSString * textData = @"aRandomTextString";
        NSString * iAsString = [NSString stringWithFormat:@"%@", array[i]];
        NSString *query = [NSString stringWithFormat:@"insert into test values('%@', %d, '%@')",
                           [textData stringByAppendingString:iAsString],
                           25+i,
                           @"someMoreRandomText"];
        [self.db executeUpdate:query];
    }
    
}

- (void)tearDown
{
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
    [self.db executeUpdate:@"DROP TABLE test"];
}

- (void)testSuccessfulCompareOrdered
{
 
    MRDatabaseContentChecker *checker = [[MRDatabaseContentChecker alloc] init];
    NSError * error;
    
    NSArray * content = @[ @[@"textData", @"numericData", @"moreText"],
                           @[@"aRandomTextString1", @(26), @"someMoreRandomText"],
                           @[@"aRandomTextString4", @(25), @"someMoreRandomText"],
                           @[@"aRandomTextString9", @(27), @"someMoreRandomText"],
                        ];
    
    BOOL dbCheckResult = [checker checkDatabase:self.db
                                          table:@"test"
                                        hasRows:content
                                        orderBy:@[@"textData"]
                                          error:&error];
    
    XCTAssertNil(error, @"An error occured checking the database");
    XCTAssertTrue(dbCheckResult, @"DB check returned false");
    
}

- (void)testSuccessfulComparedUnordered
{
    MRDatabaseContentChecker *checker = [[MRDatabaseContentChecker alloc]init];
    NSError * error;
    
    NSArray * content = @[ @[@"textData", @"numericData", @"moreText"],
                           @[@"aRandomTextString4", @(25), @"someMoreRandomText"],
                           @[@"aRandomTextString1", @(26), @"someMoreRandomText"],
                           @[@"aRandomTextString9", @(27), @"someMoreRandomText"],
                        ];
    
    BOOL dbCheckResult = [checker checkDatabase:self.db
                                          table:@"test"
                                        hasRows:content
                                          error:&error];
    
    XCTAssertNil(error, @"An error occured checking the database");
    XCTAssertTrue(dbCheckResult, @"DB check returned false");
}


@end


Classes/MRDatabaseContentChecker.m
//
//  MRDatabaseContentChecker.m
//
//  Created by Michael Rhodes on 11/05/2014.
//  Copyright (c) 2014 Mike Rhodes. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
//  except in compliance with the License. You may obtain a copy of the License at
//    http://www.apache.org/licenses/LICENSE-2.0
//  Unless required by applicable law or agreed to in writing, software distributed under the
//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
//  either express or implied. See the License for the specific language governing permissions
//  and limitations under the License.

#import "MRDatabaseContentChecker.h"

NSString* const MRDatabaseContentCheckerErrorDomain =
@"MRDatabaseContentCheckerErrorDomain";
NSString* const MRDatabaseContentCheckerErrorsArray =
@"MRDatabaseContentCheckerErrorsArray";

/**
 A simple class to check the content of your database tables.
 */
@implementation MRDatabaseContentChecker

/**
 Format the error messages produced by other methods in this
 class.
 */
- (NSString*)formattedErrors:(NSError*)error
{
    NSArray *errors = error.userInfo[MRDatabaseContentCheckerErrorsArray];
    return [errors componentsJoinedByString:@"; "];
}

/**
 Check that the contents of `table` match the specification in
 `expectedRows`. The first array in `expectedRows` MUST be the
 column names you want to check.
 
 Rows are assumed to be in the order presented in the array, and the table
 is assumed to contain the same number of rows as in the array.
 
 Note this means you can check just a subset of the data in a
 given table by including a subset of the columns.
 
 For more information on what you can have as values to check, see
 the documentation for checkDatabase:query:hasRows:error:, as this
 method calls into that one after constructing a query to get all
 the rows on the table.
 */
- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedRows
                error:(NSError* __autoreleasing *)error
{
    NSArray *columns = expectedRows[0];
    
    NSString *sql = [NSString stringWithFormat:@"select %@ from %@",
               [columns componentsJoinedByString:@", "],
               table];
    
    return [self checkDatabase:db
                         query:sql
                       hasRows:expectedRows
                         error:error];
}

/**
 Check that the contents of `table` match the specification in
 `expectedRows`. The first array in `expectedRows` MUST be the
 column names you want to check.

 Rows from the database are ordered using the coloumn 
 `orderBy` with the rows given to be in the order presented
 by the database after ordering.

 Note this means you can check just a subset of the data in a
 given table by including a subset of the columns.

 For more information on what you can have as values to check, see
 the documentation for checkDatabase:query:hasRows:error:, as this
 method calls into that one after constructing a query to get all
 the rows on the table.
 */
- (BOOL)checkDatabase:(FMDatabase*)db
                table:(NSString*)table
              hasRows:(NSArray*)expectedRows
              orderBy:(NSArray *) orderby
                error:(NSError* __autoreleasing *)error
{
    NSArray *columns = expectedRows[0];
    NSString *sql = [NSString stringWithFormat:@"select %@ from %@ order by %@",
               [columns componentsJoinedByString:@", "],
               table,
               [orderby componentsJoinedByString:@", "]];

    return [self checkDatabase:db
                         query:sql
                       hasRows:expectedRows
                         error:error];
}

/**
 Check that the result of `query` matches the specification in
 `expectedRows`. The first array in `expectedRows` MUST be the
 column names you want to check from the result set. For example:

 ```
 NSArray *expectedRows = @[
 @[@"docId",      @"revId",  @"sequenceNumber"],  // columns
 @[@"document-1", @"1-rev",  @(1)],               // expected values
 @[@"document-2", @"34-rev", @(2)]
 ];
 ```

 Expected values can be: NSString, NSNumber, NSData or NSRegularExpression.

 In the case of regular expressions, checking is based on the existence
 of one or more matches using numberOfMatchesInString:options:error:.

 Rows are assumed to be present in the result set in the order presented
 in the array, and the result set is assumed to have the same number of
 results as the array. This means that you may want to use order by or
 limit clauses in your query.

 Note this means you can check just a subset of the data in a
 given query by including a subset of the columns.
 */
- (BOOL)checkDatabase:(FMDatabase*)db
                query:(NSString*)sql
              hasRows:(NSArray*)expectedRows
                error:(NSError* __autoreleasing *)error
{
    NSArray *columns = expectedRows[0];
    NSRange range = NSMakeRange(1, [expectedRows count] - 1);
    NSArray *values = [expectedRows subarrayWithRange:range];

    FMResultSet *result = [db executeQuery:sql];

    BOOL success = YES;
    NSMutableArray *errors = [[NSMutableArray alloc] init];

    for (int i = 0; i < values.count; i++) {
        NSArray *expectedRow = values[i];

        if (![result next]) {
            success = NO;
            NSString *msg = @"Reached end of result set before array";
            [errors addObject:msg];
            break;
        }
        
        for (int col = 0; col < columns.count; col++) {
            NSString *columnName = columns[col];
            NSObject *expectedValue = expectedRow[col];

            BOOL correct;
            Class expectedClass = [expectedValue class];
            NSObject *actualValue;

            if ([expectedClass isSubclassOfClass:[NSString class]]) {
                actualValue = [result stringForColumn:columnName];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSNumber class]]) {
                long long v = [result longLongIntForColumn:columnName];
                actualValue = [NSNumber numberWithLongLong:v];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSData class]]) {
                actualValue = [result dataForColumn:columnName];
                correct = [expectedValue isEqual:actualValue];

            } else if ([expectedClass isSubclassOfClass:[NSRegularExpression class]]) {
                actualValue = [result stringForColumn:columnName];
                NSString *actualString = (NSString*)actualValue;
                NSRange range = NSMakeRange(0, actualString.length);
                NSRegularExpression *expectedRegex = (NSRegularExpression*)expectedValue;
                NSUInteger found = [expectedRegex numberOfMatchesInString:actualString
                                                                  options:0
                                                                    range:range];
                correct = (found > 0);

            } else if (expectedValue == [NSNull null]) {
                actualValue = [result objectForColumnName:columnName];
                correct = actualValue == [NSNull null];

            } else {
                // Unknown class type!
                correct = NO;
            }

            if (!correct) {
                success = NO;

                NSString *msg = [NSString stringWithFormat:
                                 @"Row %i, field %@, expected %@, actual %@",
                                 i,
                                 columnName,
                                 expectedValue,
                                 actualValue];
                [errors addObject:msg];
            }
        }
    }

    if ([result next]) {  // we should be at the end of the results
        success = NO;

        NSString *msg = @"End of array not end of result set";
        [errors addObject:msg];
    }

    [result close];

    if (errors.count > 0 && error) {
        NSString *description = NSLocalizedString(
                                                  @"Problem updating attachments table.",
                                                  nil);
        NSDictionary *userInfo = @{
                                   NSLocalizedDescriptionKey: description,
                                   MRDatabaseContentCheckerErrorsArray: [errors copy]
                                   };
        *error = [NSError errorWithDomain:MRDatabaseContentCheckerErrorDomain
                                     code:MRDatabaseContentCheckerErrorValidation
                                 userInfo:userInfo];
    }

    return success;
}

@end
